src/graph_visualization.cpp

```diff
@@ -0,0 +1,463 @@
+#include "graph_visualization.h"
+
+void initIdAndSucc(std::vector<std::vector<int>> &input, std::vector<TreeNode> &nodes);
+void predReinit(std::vector<TreeNode> &nodes);
+void degReinit(std::vector<TreeNode> &nodes);
+
+void algorithmDFS(int &exit, std::vector<TreeNode> &nodes, std::vector<bool> &usedId, TreeNode::nodeId id, TreeNode::nodeId prevId);
+void cycleExistsDFS(std::vector<TreeNode> &nodes);
+
+void heapify(std::vector<TreeNode> &nodes, int n, int i);
+void idHeapSort(std::vector<TreeNode> &nodes);
+TreeNode getTreeNodeById(std::vector<TreeNode> &nodes, TreeNode::nodeId id);
+void removeIncidentArcs(std::vector<TreeNode> &nodes, TreeNode::nodeId id);
+void greedyFAS(std::vector<TreeNode> &nodes, std::vector<std::pair<TreeNode::nodeId, TreeNode::nodeId>> &deletedEdges);
+
+void algorithmASAP(std::vector<TreeNode> &nodes);
+
+void addDummyNode(std::vector<TreeNode> &nodes, TreeNode up, TreeNode down, int order);
+void addDummyNodes(std::vector<TreeNode> &nodes, std::vector<std::pair<TreeNode::nodeId, TreeNode::nodeId>> &deletedEdges);
+
+int main() {
+  std::vector<std::vector<int>> input = {};
+  inputGraph(input);
+
+  std::vector<TreeNode> nodes = {};
+  transformationGraph(input, nodes);
+
+  printInfo(nodes);
+
+  return 0;
+}
+
+void inputGraph(std::vector<std::vector<int>> &input) {
+  for (int i = 0; 1; i++) {
+    int inputNum = -3;
+    std::vector<int> in = {};
+    input.push_back(in);
+    while (1) {
+      scanf("%i", &inputNum);
+      if (inputNum == -1 || inputNum == -2) {
+        break;
+      }
+      input[i].push_back(inputNum);
+    }
+    if (inputNum == -2) {
+      break;
+    }
+  }
+}
+
+void initIdAndSucc(std::vector<std::vector<int>> &input, std::vector<TreeNode> &nodes) {
+  for (int i = 0; i < input.size(); i++) {
+    TreeNode node = {};
+    nodes.push_back(node);
+    nodes[i].id = i;
+    for (int j = 0; j < input[i].size(); j++) {
+      nodes[i].succ.push_back(input[i][j]);
+    }
+  }
+}
+
+void predReinit(std::vector<TreeNode> &nodes) {
+  for (int i = 0; i < nodes.size(); i++) {
+    nodes[i].pred = {};
+  }
+
+  for (int i = 0; i < nodes.size(); i++) {
+    for (int j = 0; j < nodes[i].succ.size(); j++) {
+      for (int k = 0; k < nodes.size(); k++) {
+        if (nodes[k].id == nodes[i].succ[j]) {
+          nodes[k].pred.push_back(nodes[i].id);
+        }
+      }
+    }
+  }
```

Comment:
We assume that `id` is an index into `nodes`, therefore

```suggestion
  for (TreeNode &node: nodes) {
    for (TreeNode::nodeId succ: node.succ) {
      nodes[succ].pred.push_back(node.id);
    }
  }
```